### PR TITLE
Discover screen in SwiftUI

### LINKED ISSUE #587: Redesign discovery screen
Design in progress: https://www.figma.com/file/6oTsbY3vHME19RNtgSGdAw/Planetary-Design-Plan?node-id=1518%3A24052

Will do in swiftui.

### PR DESCRIPTION
**Pull request by martindsq**
Closes #587 and #1066

- [X] Implement DiscoverView
- [X] Implement a reusable InfinitGrid mimiking InfiniteList but using a Grid instead
- [X] Implement SearchResultsView reusing UniversalSearchResults (that is coupled by UIKit) as much as possible
- [X] Implement identity and message cards that can be used by the InfinitGrid
- [x] Refactor Message and Identity views to reuse code that loads the needed info from sqlite. Right now I defined CompactIdentityView and CompactPostView to use them in lists and GoldenIdentityView and GoldenPostView to use them in grids, so lets better share duplicate code to match something like:

```mermaid
graph TD
    AB[MessageButton] --> A
    A[MessageCard] --> B{Content}
    B --> |Contact|E
    B --> |Post|D{Style}
    D --> |Compact|I[CompactPostView]
    D --> |Post|K[GoldenPostView]
    EB[IdentityButton] --> E
    E[IdentityCard] --> F{Style}
    F --> |Compact|J[GoldenPostView]
    F --> |Golden|L[GoldenIdentityView]
```

Bonus:
- [X] Make the search to use profile's bios when filtering

### CODE REVIEW

File: Source/UI/Identity/GoldenIdentityView.swift [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+                colors: [Color.cardBgTop, Color.cardBgBottom],
+                startPoint: .top,
+                endPoint: .bottom
+            )
+        )
+        .cornerRadius(15)
+        .padding(EdgeInsets(top: 0, leading: 0, bottom: 0, trailing: 0))
+    }
+
+    private let goldenRatio: CGFloat = 0.618
```
Comment (mplorentz) [inline Source/UI/Identity/GoldenIdentityView.swift:76]: ah that's why it's called the GoldenIdentityView :D

File: Source/UI/Identity/IdentityButton.swift [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+//  IdentityButton.swift
+//  Planetary
+//
+//  Created by Martin Dutra on 3/1/23.
+//  Copyright © 2023 Verse Communications Inc. All rights reserved.
+//
+
+import SwiftUI
+
+struct IdentityButton: View {
```
Comment (mplorentz) [inline Source/UI/Identity/IdentityButton.swift:11]: Can you summarize what this looks like in a doc comment?

File: Source/UI/Search/SearchResultsView.swift [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+        } else {
+            async let abouts = loadAbouts(matching: normalizedQuery)
+            async let posts = loadPosts(matching: normalizedQuery)
+            return SearchResults(data: .universal(people: await abouts, posts: await posts), query: query)
+        }
+    }
+
+    func loadAbouts(matching filter: String) async -> [About] {
+        do {
+            return try await Bots.current.abouts(matching: filter)
```
Comment (mplorentz) [inline Source/UI/Search/SearchResultsView.swift:108]: Can we get the Bot from the environment rather than using the singleton here and below?

File: Source/UI/Search/SearchResultsView.swift [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+    }
+}
+
+struct SearchResultsView_Previews: PreviewProvider {
+    static var previews: some View {
+        SearchResultsView(searchText: "Hello")
+    }
+}
+
+extension Either: Identifiable, Equatable, Hashable where Left == FeedIdentifier, Right == About {
```
Comment (mplorentz) [inline Source/UI/Search/SearchResultsView.swift:224]: I think this and `SearchResults` would be better in their own files.

**Review by mplorentz — COMMENTED** [review]
This looks really nice. I left a few comments. I found myself wishing for more SwiftUIPreviews. I know we don't have a great way to add preview data to the FakeBot. Maybe we need PreviewBot that loads Feed_big.sqlite or something. But that's probably a task for the future.

File: Source/UI/Cards/CardStyle.swift [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+//
+//  Created by Martin Dutra on 12/1/23.
+//  Copyright © 2023 Verse Communications Inc. All rights reserved.
+//
+
+import Foundation
+
+enum CardStyle {
+    case compact
+    case golden
```
[the remark this answers is not in the document]
Comment (martindsq) [reply Source/UI/Cards/CardStyle.swift:17]: I added doc comments here

File: Source/UI/Discover/DiscoverView.swift [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+//  Copyright © 2022 Verse Communications Inc. All rights reserved.
+//
+
+import Analytics
+import Combine
+import CrashReporting
+import Logger
+import SwiftUI
+
+struct DiscoverView: View, HelpDrawerHost {
```
[the remark this answers is not in the document]
Comment (martindsq) [reply Source/UI/Discover/DiscoverView.swift:19]: Added.

File: Source/UI/Discover/DiscoverView.swift [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+    }
+
+    @SwiftUI.Environment(\.horizontalSizeClass)
+    var horizontalSizeClass
+
+    private let columns = [GridItem(.flexible(), spacing: 10), GridItem(.flexible())]
+
+    var body: some View {
+        ZStack(alignment: .top) {
+            if !searchText.debouncedText.isEmpty {
```
[the remark this answers is not in the document]
Comment (martindsq) [reply Source/UI/Discover/DiscoverView.swift]: You are right. I changed it to searchTextObserver.

File: Source/UI/Discover/DiscoverView.swift [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+                SearchResultsView(searchText: searchText.debouncedText)
+            } else {
+                MessageGrid(dataSource: dataSource)
+                    .placeholder(when: dataSource.isEmpty) {
+                        EmptyDiscoverView()
+                    }
+            }
+        }
+        .searchable(text: $searchText.text, placement: .navigationBarDrawer(displayMode: (.always)))
```
[the remark this answers is not in the document]
Comment (martindsq) [reply Source/UI/Discover/DiscoverView.swift]: 😢 I'm crying for all the hours I lost in the UIKit days.

File: Source/UI/Discover/DiscoverView.swift [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+    var debouncedText = ""
+
+    @Published
+    var text = ""
+
+    private var subscriptions = Set<AnyCancellable>()
+
+    init() {
+        $text
+            .removeDuplicates()
```
[the remark this answers is not in the document]
Comment (martindsq) [reply Source/UI/Discover/DiscoverView.swift:149]: I tested it and it doesn't trigger a new search, and moving removeDuplicates() to come after debounce() doesn't make a difference. I cannot explain why because combine is a bit automagical, but removeDuplicates() was called before in UniversalSearchResults (and we didn't find any issue with that) and it is called before in most examples I found in the internet. So I'm leaving as it is as it has more testing behind its shoulders.

File: Source/UI/Discover/EmptyDiscoverView.swift [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+    private var appController: AppController
+
+    var body: some View {
+        VStack(alignment: .center, spacing: 20) {
+            Image.iconPlanetary4
+            Text("Explore Planetary")
+                .font(.headline)
+                .foregroundColor(.primaryTxt)
+                .multilineTextAlignment(.center)
+            Text("The expore tab lets you see more people on Planetary. Specifically it's everything the people you follow are following.")
```
[the remark this answers is not in the document]
Comment (martindsq) [reply Source/UI/Discover/EmptyDiscoverView.swift]: Done.

File: Source/UI/Discover/EmptyDiscoverView.swift [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+
+struct EmptyDiscoverView: View {
+
+    @EnvironmentObject
+    private var appController: AppController
+
+    var body: some View {
+        VStack(alignment: .center, spacing: 20) {
+            Image.iconPlanetary4
+            Text("Explore Planetary")
```
[the remark this answers is not in the document]
Comment (martindsq) [reply Source/UI/Discover/EmptyDiscoverView.swift]: Done.

File: Source/UI/Identity/GoldenIdentityView.swift (hunk 1 above)
Comment (martindsq) [reply Source/UI/Identity/GoldenIdentityView.swift:76]: 😆

File: Source/UI/Search/SearchResultsView.swift (hunk 3 above)
Comment (martindsq) [reply Source/UI/Search/SearchResultsView.swift]: Good catch.

File: Source/UI/Search/SearchResultsView.swift (hunk 4 above)
Comment (martindsq) [reply Source/UI/Search/SearchResultsView.swift]: I moved it to the Models folder

**Review by mplorentz — APPROVED** [review]
This looks great. I see your point about changing BotRepository to just be a Bot. Let's try to prioritize that after the retreat.
